LibWeb: Separate Device and CSS pixels to allow for scaling #bigchonk

AtkinsSJ opened the pull request:
About 4 weeks after starting, here it is! And it's actually still missing some things, but I don't want to submit a PR that's any larger than this, so here we go...

Hopefully things are explained well enough in the individual commits, but some basics:

- `CSSPixels` is a float representing a value in [reference pixels](https://www.w3.org/TR/css-values-3/#reference-pixel), which are what `px` refers to in CSS.
- `DevicePixels` is an int representing a value in physical pixels on the display.
- The relationship between these is provided by `PageClient::device_pixels_per_css_pixel()`.

This allows us to render web pages the same size on regular and HiDPI displays. (And eventually also support page zooming.) Actually querying the display metrics is not implemented yet, but you can test it out with this patch:

<details>
<summary>Diff</summary>

```diff
diff --git a/Userland/Services/WebContent/PageHost.cpp b/Userland/Services/WebContent/PageHost.cpp
index 9756e1da23..2507d3a36b 100644
--- a/Userland/Services/WebContent/PageHost.cpp
+++ b/Userland/Services/WebContent/PageHost.cpp
@@ -24,6 +24,7 @@ PageHost::PageHost(ConnectionFromClient& client)
     : m_client(client)
     , m_page(make<Web::Page>(*this))
 {
+    m_screen_display_scale = 2.0f;
     setup_palette();
     m_invalidation_coalescing_timer = Web::Platform::Timer::create_single_shot(0, [this] {
         m_client.async_did_invalidate_content_rect({ m_invalidation_rect.x().value(), m_invalidation_rect.y().value(), m_invalidation_rect.width().value(), m_invalidation_rect.height().value() });
```

</details>

This touches a LOT of code throughout LibWeb, and a lot of that is interconnected in complicated ways. To try and make this not just one huge commit, most commits convert only a couple of files or classes to the new pixel units, with temporary casts where those are used. Then when those other classes are converted, the casts are also removed. (It's quite possible I've missed some though. :sweat_smile:)

Not yet following the scale factor:
- Fonts. I'm expecting this to be quite a large change on its own and this PR is already too big.
- SVG. SVG should be measured in "[User Units](https://svgwg.org/svg2-draft/coords.html#TermUserUnits)" which are then converted to CSSPixels. Again this seems like a large change.

So, those will happen in later PRs - either by myself, or by someone else who I manage to yak-bait into it. :^)

I think no changes to Ladybird are necessary to support this, but I can fix that up if it does break.

AtkinsSJ commented:
Realised after submitting: backdrop-filter is broken in this PR. I'll have a look tomorrow probably.

~~WPT wen?~~

TobyAsE commented on Userland/Libraries/LibWeb/PixelUnits.h:
```diff
@@ -0,0 +1,277 @@
+namespace Web {
+
+/// DevicePixels: A position or length on the physical display.
+class DevicePixels {
```
Isn't this just basically `using DevicePixels = float`?
I'm sure it's not

MacDue replied:
I believe `using DevicePixels = int` would just be an alias for `int`. Making it a class forces you to explicitly convert to/from `DevivePixels` (which I believe is the intention here).

Though it does seem annoying you have to write out this much boilerplate for each class.

linusg replied:
That's what `DistinctNumeric` is for

MacDue replied:
You could maybe half the boilerplate if both CSSPixels and DevicePixels inherited from some template base will all the operators. `class CSSPixels: public ArithmeticOperations<CSSPixels> {` (maybe).

AtkinsSJ replied:
I tried `DistinctNumeric` originally, but I'd wanted to keep one of the constructor private to restrict who could convert one type to the other. I ended up not doing that though.

I'll give it another try. Would be nice to not have all this boilerplate - I already had one silly typo in it that confused me for ages.

AtkinsSJ replied:
Oh, lol, I'm even still including `<AK/DistinctNumeric.h>` :laughing:

TobyAsE reviewed (commented):
What beautiful work! I tip my hat for you wanting to deal with coordinates and conversions.

Started the review, will continue :^)

MacDue commented on Userland/Libraries/LibWeb/Painting/GradientPainting.h:
```diff
@@ -24,7 +24,7 @@ using ColorStopList = Vector<ColorStop, 4>;
 struct ColorStopData {
     ColorStopList list;
-    Optional<float> repeat_length;
+    Optional<float> repeat_length; // Either degrees, or CSS Pixels.
```
This comment is not correct.
 `repeat_length` is value between 0 and 1 (it's relative to length of the gradient line, like the color stop positions)

AtkinsSJ replied:
Oops, that's an old comment from before you changed things for me!

MacDue commented on Userland/Libraries/LibWeb/Painting/FilterPainting.cpp:
```diff
@@ -98,20 +98,21 @@ void apply_filter_list(Gfx::Bitmap& target_bitmap, Layout::Node const& node, Spa
 
     // Note: The region bitmap can be smaller than the backdrop_region if it's at the edge of canvas.
-    Gfx::IntRect actual_region {};
+    DevicePixelRect actual_region {};
```
This `DevicePixelRect actual_region {}` will always be empty, `actual_region` is passed by reference to `get_region_bitmap()` and filled in by it (as an out parameter). 

This will be why filter painting is broken (as you bail out checking if `actual_region` is empty just below).

AtkinsSJ replied:
Ohhhhhh. Thanks!

MacDue reviewed (commented):
:^) Just some initial little things looking bits of my painting (not had proper look yet).
Looks like a lot of merge conflicts for me here :upside_down_face:

AtkinsSJ commented on Userland/Libraries/LibWeb/Painting/PaintableBox.cpp:
```diff
@@ -663,10 +663,10 @@ Optional<HitTestResult> PaintableBox::hit_test(Gfx::FloatPoint const& position,
 {
     if (!layout_box().children_are_inline())
-        return PaintableBox::hit_test(position, type);
+        return PaintableBox::hit_test(position.to_type<CSSPixels>(), type);
```
Good spot. I've added a check for this in https://github.com/SerenityOS/serenity/pull/16171 and built every commit here manually with that. It found a lot of other issues too. :sweat_smile:

AtkinsSJ commented on Userland/Services/WebContent/PageHost.h:
```diff
@@ -85,8 +87,9 @@ class PageHost final : public Web::PageClient {
-    Gfx::IntSize m_content_size;
+    Web::DevicePixelRect m_screen_rect;
+    Web::DevicePixelSize m_content_size;
+    float m_screen_display_scale { 1.0f };
```
I decided to add a setter instead.

AtkinsSJ commented on Userland/Services/WebContent/PageHost.cpp:
```diff
@@ -104,10 +104,12 @@ Web::Layout::InitialContainingBlock* PageHost::layout_root()
-void PageHost::paint(Gfx::IntRect const& content_rect, Gfx::Bitmap& target)
+void PageHost::paint(Web::DevicePixelRect const& content_rect, Gfx::Bitmap& target)
 {
+    Gfx::IntRect int_content_rect { content_rect.x().value(), content_rect.y().value(), content_rect.width().value(), content_rect.height().value() };
```
Realised a later commit replaced this with some `to_type<int>()` calls, so I've done that here.

AtkinsSJ commented:
So....
- Rebased and fixed lots of merge conflicts.
- Rebuilt every single commit individually, using https://github.com/SerenityOS/serenity/pull/16171 - this turned up a bunch of `to_type<>()`s that weren't useful, and other build issues from previous rebases. This took way too long. :smiling_face_with_tear: 
- Applied feedback.
- Added an experimental commit converting the DevicePixels and CSSPixels types to DistinctNumeric. It probably is an improvement, though there are some small disadvantages. My plan is to see what people think of that, then squash those changes into the correct commits.

AtkinsSJ commented:
Future plan:
- Either integrate or discard the DistinctNumeric changes.
- Split this PR into chunks, instead of expecting people to review the whole thing at once.

MacDue commented on Userland/Libraries/LibWeb/Painting/GradientPainting.cpp:
```diff
@@ -257,7 +257,7 @@ void paint_linear_gradient(PaintContext& context, DevicePixelRect const& gradien
     DevicePixelPoint offset { cos_angle * (gradient_length_px / 2), sin_angle * (gradient_length_px / 2) };
     auto center = gradient_rect.translated(-gradient_rect.location()).center();
-    auto start_point = center.to_type<float>() - offset.to_type<float>();
+    auto start_point = center.to_type<int>().to_type<float>() - offset.to_type<int>().to_type<float>();
```
What's with the `.to_type<int>().to_type<float>()`? I've seen this in a few places, seems a little confusing.

MacDue replied:
(Maybe) we could add our `cast` function to LibWeb that could handle the special cases like this. So this could be just something like `cast_to<float>(center)`

AtkinsSJ replied:
I've gone with adding `IntPoint` overloads to the Painter methods that are used. See: https://github.com/SerenityOS/serenity/pull/16366

AtkinsSJ commented:
Marking this as a draft - finally starting to split it up and submit the pieces as their own PRs, so it's less of a beast to review. Part 1: https://github.com/SerenityOS/serenity/pull/16366
